src/events.ts

```diff
@@ -0,0 +1,81 @@
+import { MraidState } from "./state";
+
+export enum MraidEvent {
+  Ready = "ready",
+  Error = "error",
+  StateChange = "stateChange",
+  ViewableChange = "viewableChange",
+}
+
+export type EventListener<T1, T2> = (t1: T1, t2: T2) => void;
+export type ErrorEventListener = EventListener<string, string | null>;
+export type StateChangeEventListener = EventListener<MraidState, void>;
+export type ViewableChangeEventListener = EventListener<boolean, void>;
+export type ReadyEventListener = EventListener<void, void>;
+
+export type MraidEventListener =
+  | ErrorEventListener
+  | StateChangeEventListener
+  | ViewableChangeEventListener
+  | ReadyEventListener
+  | null
+  | undefined;
+
+export class EventsCoordinator {
+  private eventListeners: Map<string, Set<MraidEventListener>> = new Map();
+
+  addEventListener(event: MraidEvent, listener: MraidEventListener) {
+    if (!this.isCorrectEvent(event)) {
+      // log error
+      return;
+    }
+
+    if (listener) {
+      const listeners = this.eventListeners.get(event) ?? new Set();
+      listeners.add(listener);
+      this.eventListeners.set(event, listeners);
+    }
+  }
+
+  removeEventListener(event: MraidEvent, listener: MraidEventListener) {
+    if (!this.isCorrectEvent(event)) {
+      // log error
+      return;
+    }
+
+    if (listener) {
+      const listeners = this.eventListeners.get(event)?.delete(listener);
+    } else {
+      this.eventListeners.delete(event);
+    }
+  }
+
+  fireReadyEvent() {
+    this.eventListeners.get(MraidEvent.Ready)?.forEach((value) => {
+      (value as ReadyEventListener)?.();
+    });
+  }
+
+  fireErrorEvent(message: string, action: string | null) {
+    this.eventListeners.get(MraidEvent.Error)?.forEach((value) => {
+      (value as ErrorEventListener)?.(message, action);
+    });
+  }
+
+  fireStateChangeEvent(newState: MraidState) {
+    this.eventListeners.get(MraidEvent.StateChange)?.forEach((value) => {
+      (value as StateChangeEventListener)?.(newState);
+    });
+  }
+
+  fireViewableChangeEvent(isViewable: boolean) {
+    this.eventListeners.get(MraidEvent.ViewableChange)?.forEach((value) => {
+      (value as ViewableChangeEventListener)?.(isViewable);
+    });
+  }
+
+  private isCorrectEvent(event: MraidEvent): boolean {
+    // TODO: Object.values is ES2017. Check if it will work for Android and iOS
```

Comment:
Don't hesitate to create an appropriate ticket for checking this assumption on both platform. Maybe you won't forget yourself to test this on Android, but for iOS, I'm not quite sure.